htp/htp_transaction.c

```diff
@@ -574,22 +823,55 @@ htp_status_t htp_tx_req_process_body_data_ex(htp_tx_t *tx, const void *data, siz
     // NULL data is allowed in this private function; it's
     // used to indicate the end of request body.
 
-    // Keep track of the body length.
-    tx->request_entity_len += len;
-
     // Send data to the callbacks.
 
     htp_tx_data_t d;
     d.tx = tx;
     d.data = (unsigned char *) data;
     d.len = len;
 
-    htp_status_t rc = htp_req_run_hook_body_data(tx->connp, &d);
-    if (rc != HTP_OK) {
-        htp_log(tx->connp, HTP_LOG_MARK, HTP_LOG_ERROR, 0, "Request body data callback returned error (%d)", rc);
-        return HTP_ERROR;
+    enum htp_content_encoding_t encoding_processing;
+    if (tx->connp->cfg->response_decompression_enabled) {
+        encoding_processing = tx->request_content_encoding;
+    } else {
+        encoding_processing = HTP_COMPRESSION_NONE;
     }
+    switch (encoding_processing) {
+        case HTP_COMPRESSION_GZIP:
+        case HTP_COMPRESSION_DEFLATE:
+        case HTP_COMPRESSION_LZMA:
+            // In severe memory stress these could be NULL
+            if (tx->connp->out_decompressor == NULL || tx->connp->out_decompressor->decompress == NULL)
+                return HTP_ERROR;
+
+            // Send data buffer to the decompressor.
+            tx->connp->out_decompressor->decompress(tx->connp->out_decompressor, &d);
 
+            if (data == NULL) {
+                // Shut down the decompressor, if we used one.
+                htp_connp_destroy_decompressors(tx->connp);
+            }
+            break;
+
+        case HTP_COMPRESSION_NONE:
```

Comment:
think this should be the first entry as it is most likely to be true

Reply:
ok